File: src/main/java/com/github/pareronia/i18n_puzzles/Puzzle2025_10.java
package com.github.pareronia.i18n_puzzles;

import static java.util.stream.Collectors.toMap;

import java.nio.charset.StandardCharsets;
import java.text.Normalizer;
import java.text.Normalizer.Form;
import java.util.ArrayDeque;
import java.util.Deque;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;
import java.util.concurrent.ConcurrentHashMap;

import org.springframework.security.crypto.bcrypt.BCrypt;

import com.github.pareronia.i18n_puzzles.common.Sample;
import com.github.pareronia.i18n_puzzles.common.Samples;
import com.github.pareronia.i18n_puzzles.common.SolutionBase;
import com.github.pareronia.i18n_puzzles.common.Utils;

public class Puzzle2025_10 extends SolutionBase<Integer>{

    private final Map<Integer, Boolean> cache = new ConcurrentHashMap<>();

    private Puzzle2025_10(final boolean debug) {
        super(debug);
    }

    public static Puzzle2025_10 create() {
        return new Puzzle2025_10(false);
    }

    public static Puzzle2025_10 createDebug() {
        return new Puzzle2025_10(true);
    }

    private Set<String> variants(final String secret) {
        record State(int i, String var) {}
        final Set<String> ans = new HashSet<>();
        final Deque<State> q = new ArrayDeque<>();
        q.add(new State(0, ""));
        while (!q.isEmpty()) {
            final State state = q.poll();
            if (state.i == secret.length()) {
                ans.add(state.var);
            } else {
                final String c = String.valueOf(secret.charAt(state.i));
                final String n = Normalizer.normalize(c, Form.NFD);
                if (!c.equals(n)) {
                    q.add(new State(state.i + 1, state.var + n));
                }
                q.add(new State(state.i + 1, state.var + c));
            }
        }
        return ans;
    }

    private boolean check(final String secret, final String hash) {
        return cache.computeIfAbsent(
            Objects.hash(secret, hash),
            k -> variants(secret).stream().anyMatch(s ->
                BCrypt.checkpw(s.getBytes(StandardCharsets.UTF_8), hash)));
    }

    @Override
    public Integer solve(final List<String> input) {
        final List<List<String>> blocks = Utils.toBlocks(input);
        final Map<String, String> auths = blocks.get(0).stream()
                .map(line -> line.split(" "))
                .collect(toMap(split -> split[0], split -> split[1]));
        return (int) blocks.get(1).parallelStream()
                .map(line -> line.split(" "))
                .filter(split -> {
                    final String username = split[0];
                    final String cand = Normalizer.normalize(split[1], Form.NFC);
                    return check(cand, auths.get(username));
                })
                .count();
    }

    @Override
    @Samples({
        @Sample(input = TEST, expected = "4")
    })
    public void samples() { }

    public static void main(final String[] args) throws Exception {
        Puzzle2025_10.create().run();
    }

    private static final String TEST = """
            etasche $2b$07$0EBrxS4iHy/aHAhqbX/ao.n7305WlMoEpHd42aGKsG21wlktUQtNu
            mpataki $2b$07$bVWtf3J7xLm5KfxMLOFLiu8Mq64jVhBfsAwPf8/xx4oc5aGBIIHxO
            ssatterfield $2b$07$MhVCvV3kZFr/Fbr/WCzuFOy./qPTyTVXrba/2XErj4EP3gdihyrum
            mvanvliet $2b$07$gf8oQwMqunzdg3aRhktAAeU721ZWgGJ9ZkQToeVw.GbUlJ4rWNBnS
            vbakos $2b$07$UYLaM1I0Hy/aHAhqbX/ao.c.VkkUaUYiKdBJW5PMuYyn5DJvn5C.W
            ltowne $2b$07$4F7o9sxNeaPe..........l1ZfgXdJdYtpfyyUYXN/HQA1lhpuldO

            etasche .pM?XÑ0i7ÈÌ
            mpataki 2ö$p3ÄÌgÁüy
            ltowne 3+sÍkÜLg._
            ltowne 3+sÍkÜLg?_
            mvanvliet *íÀŸä3hñ6À
            ssatterfield 8É2U53N~Ë
            mpataki 2ö$p3ÄÌgÁüy
            mvanvliet *íÀŸä3hñ6À
            etasche .pM?XÑ0i7ÈÌ
            ssatterfield 8É2U53L~Ë
            mpataki 2ö$p3ÄÌgÁüy
            vbakos 1F2£èÓL
            """;
}

File: src/main/python/Puzzle2025_10.py
import sys
import unicodedata
from collections import deque
from functools import cache

import bcrypt
from i18n_puzzles.common import InputData
from i18n_puzzles.common import SolutionBase
from i18n_puzzles.common import i18n_puzzles_samples
from i18n_puzzles.common import to_blocks

TEST = """\
etasche $2b$07$0EBrxS4iHy/aHAhqbX/ao.n7305WlMoEpHd42aGKsG21wlktUQtNu
mpataki $2b$07$bVWtf3J7xLm5KfxMLOFLiu8Mq64jVhBfsAwPf8/xx4oc5aGBIIHxO
ssatterfield $2b$07$MhVCvV3kZFr/Fbr/WCzuFOy./qPTyTVXrba/2XErj4EP3gdihyrum
mvanvliet $2b$07$gf8oQwMqunzdg3aRhktAAeU721ZWgGJ9ZkQToeVw.GbUlJ4rWNBnS
vbakos $2b$07$UYLaM1I0Hy/aHAhqbX/ao.c.VkkUaUYiKdBJW5PMuYyn5DJvn5C.W
ltowne $2b$07$4F7o9sxNeaPe..........l1ZfgXdJdYtpfyyUYXN/HQA1lhpuldO

etasche .pM?XÑ0i7ÈÌ
mpataki 2ö$p3ÄÌgÁüy
ltowne 3+sÍkÜLg._
ltowne 3+sÍkÜLg?_
mvanvliet *íÀŸä3hñ6À
ssatterfield 8É2U53N~Ë
mpataki 2ö$p3ÄÌgÁüy
mvanvliet *íÀŸä3hñ6À
etasche .pM?XÑ0i7ÈÌ
ssatterfield 8É2U53L~Ë
mpataki 2ö$p3ÄÌgÁüy
vbakos 1F2£èÓL
"""

Output = int


class Solution(SolutionBase[Output]):
    def solve(self, input: InputData) -> Output:
        @cache
        def check(secret: str, hash: str) -> bool:
            def variants(secret: str) -> set[str]:
                ans = set[str]()
                q = deque[tuple[int, str]]()
                q.append((0, ""))
                while q:
                    size, var = q.popleft()
                    if size == len(secret):
                        ans.add(var)
                    else:
                        c = secret[size]
                        n = unicodedata.normalize("NFD", c)
                        if n != c:
                            q.append((size + 1, var + n))
                        q.append((size + 1, var + c))
                return ans

            hashed = hash.encode()
            return any(
                bcrypt.checkpw(s.encode(), hashed) for s in variants(secret)
            )

        blocks = to_blocks(input)
        auths = {a: b for a, b in map(lambda line: line.split(), blocks[0])}
        return sum(
            check(unicodedata.normalize("NFC", attempt), auths[username])
            for username, attempt in map(lambda line: line.split(), blocks[1])
        )

    @i18n_puzzles_samples(((TEST, 4),))
    def samples(self) -> None:
        pass


solution = Solution(10)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()
